File: tools/upstream-sync/frontmatter.go
package main

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

func sanitizeSkillFrontmatter(root string, fields []string) error {
	if len(fields) == 0 {
		return nil
	}
	return filepath.WalkDir(root, func(path string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return fmt.Errorf("walk %s: %w", path, walkErr)
		}
		if entry.IsDir() || entry.Name() != "SKILL.md" {
			return nil
		}
		if err := removeFrontmatterFields(path, fields); err != nil {
			return fmt.Errorf("sanitize %s: %w", path, err)
		}
		return nil
	})
}

func removeFrontmatterFields(path string, fields []string) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("read skill: %w", err)
	}
	lines := strings.SplitAfter(string(data), "\n")
	if len(lines) < 2 || strings.TrimSpace(lines[0]) != "---" {
		return fmt.Errorf("missing YAML frontmatter")
	}
	fieldSet := make(map[string]struct{}, len(fields))
	for _, field := range fields {
		fieldSet[field] = struct{}{}
	}

	frontmatterClosed := false
	output := make([]string, 0, len(lines))
	output = append(output, lines[0])
	for _, line := range lines[1:] {
		trimmed := strings.TrimSpace(line)
		if !frontmatterClosed && trimmed == "---" {
			frontmatterClosed = true
			output = append(output, line)
			continue
		}
		if !frontmatterClosed && hasFrontmatterField(line, fieldSet) {
			continue
		}
		output = append(output, line)
	}
	if !frontmatterClosed {
		return fmt.Errorf("unterminated YAML frontmatter")
	}
	return writeFileAtomic(path, []byte(strings.Join(output, "")))
}

func hasFrontmatterField(line string, fields map[string]struct{}) bool {
	if len(line) == 0 || line[0] == ' ' || line[0] == '\t' {
		return false
	}
	name, _, found := strings.Cut(strings.TrimRight(line, "\r\n"), ":")
	if !found {
		return false
	}
	_, exists := fields[name]
	return exists
}

File: tools/upstream-sync/src/frontmatter.ts
import { readdir, readFile, stat, writeFile } from 'node:fs/promises';
import { join } from 'node:path';

class FrontmatterError extends Error {
  name = 'FrontmatterError';
}

export async function sanitizeSkills(root: string, fields: string[]): Promise<void> {
  if (fields.length === 0) return;
  const info = await stat(root);
  if (!info.isDirectory()) return;
  await visit(root, new Set(fields));
}

async function visit(directory: string, fields: Set<string>): Promise<void> {
  const entries = await readdir(directory, { withFileTypes: true });
  for (const entry of entries) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) {
      await visit(path, fields);
      continue;
    }
    if (entry.name !== 'SKILL.md') continue;
    const source = await readFile(path, 'utf8');
    await writeFile(path, removeFrontmatterFields(source, fields));
  }
}

function removeFrontmatterFields(source: string, fields: Set<string>): string {
  const lines = source.match(/.*(?:\r\n|\n|$)/g)?.filter((line) => line.length > 0) ?? [];
  if (lines.length < 2 || lines[0]?.trim() !== '---') throw new FrontmatterError('missing YAML frontmatter');
  let closed = false;
  const output: string[] = [lines[0]];
  for (const line of lines.slice(1)) {
    const trimmed = line.trim();
    if (!closed && trimmed === '---') {
      closed = true;
      output.push(line);
      continue;
    }
    if (!closed && isRemovedField(line, fields)) continue;
    output.push(line);
  }
  if (!closed) throw new FrontmatterError('unterminated YAML frontmatter');
  return output.join('');
}

function isRemovedField(line: string, fields: Set<string>): boolean {
  if (line.startsWith(' ') || line.startsWith('\t')) return false;
  const separator = line.indexOf(':');
  return separator >= 0 && fields.has(line.slice(0, separator));
}
